ddtrace/internal/products.py

```diff
@@ -0,0 +1,178 @@
+import atexit
+from collections import defaultdict
+from collections import deque
+import sys
+import typing as t
+
+from ddtrace.internal import forksafe
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.uwsgi import check_uwsgi
+from ddtrace.internal.uwsgi import uWSGIConfigError
+from ddtrace.internal.uwsgi import uWSGIMasterProcess
+
+
+log = get_logger(__name__)
+
+if sys.version_info < (3, 10):
+    from importlib_metadata import entry_points
+else:
+    from importlib.metadata import entry_points
+
+try:
+    from typing import Protocol  # noqa:F401
+except ImportError:
+    from typing_extensions import Protocol  # type: ignore[assignment]
+
+
+class Product(Protocol):
+    requires: t.List[str]
+
+    def post_preload(self) -> None:
+        ...
+
+    def start(self) -> None:
+        ...
+
+    def restart(self, join: bool = False) -> None:
+        ...
+
+    def stop(self, join: bool = False) -> None:
+        ...
+
+    def at_exit(self, join: bool = False) -> None:
+        ...
+
+
+class ProductManager:
+    __products__: t.Dict[str, Product] = {}  # All discovered products
+
+    def __init__(self) -> None:
+        self._products: t.Optional[t.List[t.Tuple[str, Product]]] = None  # Topologically sorted products
+
+        for product_plugin in entry_points(group="ddtrace.products"):
```

Comment:
So when we start implementing products, would we make them discoverable via `entry_points(group="ddtrace.products")` by adding them to the `pyproject.toml` file (like you did [here](https://github.com/DataDog/dd-trace-py/pull/10173/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711))?

Reply:
Yes! 🙂